src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -773,7 +773,9 @@ public override ForEachStatementInfo GetForEachStatementInfo(ForEachStatementSyn
 
         public override ForEachStatementInfo GetForEachStatementInfo(CommonForEachStatementSyntax node)
         {
-            BoundForEachStatement boundForEach = (BoundForEachStatement)GetUpperBoundNode(node);
+            // Declaration expressions can result in having a BoundBlock associated with a given foreach statement,
+            // in addition to a BoundForEachStatement
+            BoundForEachStatement boundForEach = GetBoundNodes(node).OfType<BoundForEachStatement>().FirstOrDefault();
```

Comment:
Thanks for the clarification. `WrapWithVariablesIfAny` indeed marks the `BoundBlock` as compiler generated, so I can use that flag.